File: svm/svm.jl
using CSV
using DataFrames
using MLJ
using MLJScikitLearnInterface
using MLJBase: machine, fit!, transform, coerce
using DataFrames: DataFrame, dropmissing

function preprocess(data::DataFrame)
    # Extract target column
    y = data.Class
    data = data[:, Not(:Class)]
  
    # Normalize numerical columns
    stand = Standardizer(count=true)
    X = transform(fit!(machine(stand, data)), data)

    # Coerce the target column to a multiclass type
    y = coerce(y, Multiclass)
    
    return X, y
end

function main()
    # Generate and print random seed
    rng = rand(1:1000)

    # Load data
    data = CSV.File("../data/creditcard_2023.csv") |> DataFrame
    X, y = preprocess(data)

    # Split the data, 80% for training and 20% for testing, shuffles the data
    (X_train, X_test), (y_train, y_test) = partition((X, y), 0.8, rng = rng, multi=true)

    # Load the model
    SVMLinearClassifier = @load SVMLinearClassifier pkg=MLJScikitLearnInterface
    model = SVMLinearClassifier(random_state=rng)

    # Train the model
    mach = machine(model, X_train, y_train)
    fit!(mach)

    # Make predictions
    ŷ = predict(mach, X_test)

    #Calculate accuracy
    accuracy = mean(ŷ.== y_test)
    println("Random Seed: ", rng)
    println("Accuracy: $accuracy")
end

main()

File: svm/svm.py
import pandas as pd
import random
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn import svm

def preprocess(data):

    # Identify numerical columns excluding the target column
    numerical_cols = data.drop(columns=['Class']).select_dtypes(include=['int64', 'float64']).columns

    # Scale numerical columns
    scaler = StandardScaler()
    data[numerical_cols] = scaler.fit_transform(data[numerical_cols])
    
    return data


def main():
    # Set random seed for reproducibility
    randint = random.randint(0, 1000) 

    # Load dataset
    data = pd.read_csv('../data/creditcard_2023.csv')
    target_col = 'Class'

    # Preprocess data
    data = preprocess(data)

    # Split features and target
    X = data.drop(columns=[target_col])  # Features
    y = data[target_col]  # Target variable
    
    # Train-test split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=randint)
    
    # Load and train the model
    clf = svm.LinearSVC(random_state=randint)
    clf = clf.fit(X_train, y_train)

    # Make predictions
    y_pred = clf.predict(X_test)

    # Evaluate accuracy
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Random Seed: {randint}")
    print(f"Accuracy: {accuracy}")

if __name__ == "__main__":
    main()

